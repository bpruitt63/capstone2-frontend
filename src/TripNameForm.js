import React, {useState} from 'react';
import {useHandleChange, useErrors} from './hooks';
import BoateyApi from './BoateyApi';
import Errors from './Errors';

/** Form to name a trip and save trip to database */
function TripNameForm({currentTrip, username, deleteCurrentTrip}) {

    const [data, handleChange, setData] = useHandleChange({name: ''});
    const [apiErrors, getApiErrors, setApiErrors] = useErrors();
    const [formErrors, setFormErrors] = useState({});

    const onSubmit = async (e) => {
        e.preventDefault();
        setApiErrors({});

        /** Validates form data */
        if (!data.name || data.name.length < 1 || data.name.length > 30) {
            setFormErrors({error: 'Trip Name must be between one and thirty characters.'});
            return false;
        } else {
            setFormErrors({});
            try{

                /** Creates trip object and submits to database */
                let locations = [];
                for (let location of currentTrip) {
                    locations.push({location: location.id});
                };
                const tripData = {trip: {name: data.name,
                                        distance: currentTrip[currentTrip.length - 1].totalDistance},
                                        locations}
                await BoateyApi.addTrip(username, tripData);

                /** Removes trip data from state once it has been submitted */
                setData({name: ''});
                deleteCurrentTrip();
            } catch (err) {
                getApiErrors(err);
            };
        };
    };

    return (
        <>
            <Errors formErrors={formErrors}
                    apiErrors={apiErrors} />
            <form onSubmit={onSubmit}>
                <label htmlFor='name'>Name Trip: </label>
                <input type='text'
                        name='name'
                        placeholder='Trip Name'
                        value={data.name}
                        onChange={handleChange} />
                <button>Save Trip</button>
            </form>
        </>
    )
};

export default TripNameForm;