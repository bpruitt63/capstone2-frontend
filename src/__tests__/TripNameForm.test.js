import React from 'react';
import { render, screen } from '@testing-library/react';
import TripNameForm from '../TripNameForm';

test('renders', () => {
    render(<TripNameForm/>);
    expect(screen.getByText(/Save Trip/)).toBeInTheDocument;
});

test('matches snapshot', () => {
    const {asFragment} = render(<TripNameForm/>);
    expect(asFragment()).toMatchSnapshot();
});