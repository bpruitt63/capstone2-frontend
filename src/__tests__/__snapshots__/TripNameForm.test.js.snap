// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Errors"
  />
  <form
    class="tripNameForm"
  >
    <div
      class="input-group"
    >
      <input
        class="form-control"
        name="name"
        placeholder="Trip Name"
        type="text"
        value=""
      />
      <div
        class="input-group-append"
      >
        <button
          class="btn btn-secondary"
        >
          Save Trip
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
